File: module_3/module_3_python/personne_factory.py
"""
Pattern de création: Factory Method

PersonneFactory : crée des objets Etudiant ou Enseignant
sans exposer la logique d'instanciation au client.

"""

from etudiant import Etudiant
from enseignant import Enseignant


class PersonneFactory:
    """
    Fabrique de personnes.
    La méthode de classe creer() joue le rôle de factory method.
    """

    @classmethod
    def creer(cls, type_personne: str, **kwargs):
        """
        Crée et retourne une instance selon le type demandé.

        """
        type_lower = type_personne.strip().lower()

        if type_lower == "etudiant":
            champs = ("nom", "age", "numero_etudiant", "moyenne")
            cls.__verifier_champs(champs, kwargs, "Etudiant")
            return Etudiant(
                nom=kwargs["nom"],
                age=kwargs["age"],
                numero_etudiant=kwargs["numero_etudiant"],
                moyenne=kwargs["moyenne"],
            )

        elif type_lower == "enseignant":
            champs = ("nom", "age", "matiere", "salaire")
            cls.__verifier_champs(champs, kwargs, "Enseignant")
            return Enseignant(
                nom=kwargs["nom"],
                age=kwargs["age"],
                matiere=kwargs["matiere"],
                salaire=kwargs["salaire"],
            )

        else:
            raise ValueError(
                f"Type inconnu : '{type_personne}'. "
                "Valeurs acceptées : 'etudiant', 'enseignant'."
            )

    @staticmethod
    def __verifier_champs(champs_requis, kwargs, nom_classe):
        manquants = [c for c in champs_requis if c not in kwargs]
        if manquants:
            raise ValueError(
                f"Champs manquants pour créer un {nom_classe} : {manquants}"
            )

File: module_3/Module_3_Java/src/main/java/com/ensta/gestion/factory/PersonneFactory.java
package com.ensta.gestion.factory;

import com.ensta.gestion.modele.Etudiant;
import com.ensta.gestion.modele.Personne;
import com.ensta.gestion.modele.Enseignant;



/**
 * PATTERN 2 : FACTORY METHOD
 * Centralise la logique de création des Personne (Etudiant, Enseignant).
 * Le client n'utilise jamais directement "new Etudiant(...)" ou "new Enseignant(...)".
 *
 * Principe SOLID respecté :
 *  - Single Responsibility : la factory est seule responsable de l'instanciation.
 *  - Open/Closed : ajouter un nouveau type (ex. Administrateur) ne modifie pas
 *    les méthodes existantes.
 */
public class PersonneFactory {

    public enum TypePersonne { ETUDIANT, ENSEIGNANT }

    /**
     * Crée et retourne une Personne selon le type demandé.
     *
     * @param type            ETUDIANT ou ENSEIGNANT
     * @param nom             Nom de la personne
     * @param age             Âge de la personne
     * @param infoComplementaire  Numéro étudiant  OU  matière enseignée
     * @return instance de Etudiant ou Enseignant
     */
    public static Personne creer(TypePersonne type, String nom, int age,
                                 String infoComplementaire) {
        return switch (type) {
            case ETUDIANT   -> new Etudiant(nom, age, infoComplementaire);
            case ENSEIGNANT -> new Enseignant(nom, age, infoComplementaire);
        };
    }

    /** Raccourci pour créer un Etudiant directement typé. */
    public static Etudiant creerEtudiant(String nom, int age, String numero) {
        return (Etudiant) creer(TypePersonne.ETUDIANT, nom, age, numero);
    }

    /** Raccourci pour créer un Enseignant directement typé. */
    public static Enseignant creerEnseignant(String nom, int age, String matiere) {
        return (Enseignant) creer(TypePersonne.ENSEIGNANT, nom, age, matiere);
    }
}